### PR TITLE
suggested amendments for PR #2935 (round 2)

### PR DESCRIPTION
**Pull request by AlexHentschel**
• making BlockOrDelayFunc concurrency safe
• adding documentation
• fixing numerical edge cases

### CODE REVIEW

File: consensus/integration/blockordelay_test.go [hunk 1]
```diff
@@ -38,19 +44,45 @@ func blockNodesFirstMessages(n uint64, denyList ...*Node) BlockOrDelayFunc {
...
-		block := rand.Intn(100) <= percent
+		lock.Lock()
+		block := prng.Float32() < dropProbability
+		lock.Unlock()
 		return block, 0
 	}
 }
 
+// delayReceiverMessagesByRange delivers all messages, but with a randomly sampled
+// delay in the interval [low, high). Panicks
```
Comment (jordanschalm) [inline consensus/integration/blockordelay_test.go:60]:
```suggestion
// delay in the interval [low, high). Panics when `low` < 0 or `low > `high`
```

File: consensus/integration/integration_test.go [hunk 2]
```diff
@@ -137,27 +137,35 @@ func chainViews(t *testing.T, node *Node) []uint64 {
...
-// block nothing
-func blockNothing(channel network.Channel, event interface{}, sender, receiver *Node) (bool, time.Duration) {
+// blockNothing specifies that _all_ messages should be delivered without delay.
+// I.e. this function returns always `false` (no blocking), `0` (no delay).
+func blockNothing(_ network.Channel, _ interface{}, _, _ *Node) (bool, time.Duration) {
 	return false, 0
 }
 
-// block all messages sent by or received by a list of denied nodes
+// blockNodes specifies that all messages sent by or received any member of the `denyList`
```
Comment (jordanschalm) [inline consensus/integration/integration_test.go:153]:
```suggestion
// blockNodes specifies that all messages sent or received by any member of the `denyList`
```

File: consensus/integration/blockordelay_test.go [hunk 3]
```diff
@@ -38,19 +44,45 @@ func blockNodesFirstMessages(n uint64, denyList ...*Node) BlockOrDelayFunc {
 	}
 }
 
-func blockReceiverMessagesByPercentage(percent int) BlockOrDelayFunc {
-	rand.Seed(time.Now().UnixNano())
+// blockReceiverMessagesRandomly drops messages randomly with a probability of `dropProbability` ∈ [0,1]
+func blockReceiverMessagesRandomly(dropProbability float32) BlockOrDelayFunc {
+	lock := new(sync.Mutex)
+	prng := rand.New(rand.NewSource(64))
```
Comment (jordanschalm) [inline consensus/integration/blockordelay_test.go:50]: Why 64? 🤔

**Review by jordanschalm — APPROVED** [review]

File: consensus/integration/blockordelay_test.go (hunk 3 above)
Comment (AlexHentschel) [reply consensus/integration/blockordelay_test.go]: 🤦 that is an artifact of my experimenting with initializing the random number generator. Fixed in https://github.com/onflow/flow-go/pull/3096/commits/2533b3e2aba7cd3b391781f21adb141e5a3e8ed3

**Pull request merged**